File: C++/queue.cpp
#include <stdio.h>
#include <semaphore.h>
#include <thread>
#include <vector>

using std::vector, std::thread;

sem_t leaderQueue;
sem_t followerQueue;
sem_t mutex;
int counter = 0;
int numthreads = 10;
vector<thread> threads;

void leader() {
    sem_post(&followerQueue);
    sem_wait(&leaderQueue);

    sem_wait(&mutex);
    counter++;
    printf("count: %d\n", counter);
    sem_post(&mutex);
}

void follower() {
    sem_post(&leaderQueue);
    sem_wait(&followerQueue);

    sem_wait(&mutex);
    counter--;
    printf("count: %d\n", counter);
    sem_post(&mutex);
}

int main() {
    sem_init(&leaderQueue, 0, 0);
    sem_init(&followerQueue, 0, 0);
    sem_init(&mutex, 0, 1);

    for (int i = 0; i < numthreads; ++i) {
        threads.push_back(thread(leader));
    }

    for (int i = 0; i < numthreads; ++i) {
        threads.push_back(thread(follower));
    }

    for (thread &t : threads) {
        t.join();
    }

    sem_destroy(&leaderQueue);
    sem_destroy(&followerQueue);
    sem_destroy(&mutex);

    return 0;
}

File: python/queue.py
from threading import *

leaderQueue = Semaphore(0)
followerQueue = Semaphore(0)
mutex = Semaphore(1)
count = 0
threads = []
numthreads = 10

def leader():
    followerQueue.release()
    leaderQueue.acquire()

    mutex.acquire()
    global count
    count += 1
    print(f"count: {count}\n")
    mutex.release()

def follower():
    leaderQueue.release()
    followerQueue.acquire()

    mutex.acquire()
    global count

    count -= 1
    print(f"count: {count}\n")
    mutex.release()

for i in range(numthreads):
    threads.append(Thread(target=leader))

for i in range(numthreads):
    threads.append(Thread(target=follower))

for thread in threads:
    thread.start()
    
for thread in threads:
    thread.join()

print("All threads have finished.")
